prow/plank/reconciler.go

```diff
@@ -0,0 +1,645 @@
+	log                *logrus.Entry
+	config             config.Getter
+	totURL             string
+	clock              clock.Clock
+}
+
+func (r *reconciler) syncMetrics(stop <-chan struct{}) error {
+	for {
+		select {
+		case <-stop:
+			return nil
+		case <-time.NewTicker(30 * time.Second).C:
+			pjs := &prowv1.ProwJobList{}
+			if err := r.pjClient.List(r.ctx, pjs, optAllProwJobs()); err != nil {
+				r.log.WithError(err).Error("failed to list prowjobs for metrics")
+				continue
+			}
+			kube.GatherProwJobMetrics(pjs.Items)
+		}
+	}
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	if r.overwriteReconcile != nil {
+		return r.overwriteReconcile(request)
+	}
+	return r.defaultReconcile(request)
+}
+
+func (r *reconciler) defaultReconcile(request reconcile.Request) (reconcile.Result, error) {
+	pj := &prowv1.ProwJob{}
+	if err := r.pjClient.Get(r.ctx, request.NamespacedName, pj); err != nil {
+		if !kerrors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("failed to get prowjob %s: %v", request.Name, err)
+		}
+
+		// Objects can be deleted from the API while being in our workqueue
+		return reconcile.Result{}, nil
+	}
+
+	// TODO: Terminal errors for unfixable cases like missing build clusters
+	// and not return an error to prevent requeuing?
+	res, err := r.reconcile(pj)
+	if res == nil {
+		res = &reconcile.Result{}
+	}
+	if err != nil {
+		r.log.WithError(err).Error("Reconciliation failed")
+	}
+	return *res, err
+}
+
+func (r *reconciler) reconcile(pj *prowv1.ProwJob) (*reconcile.Result, error) {
+	// terminateDupes first, as that might reduce cluster load and prevent us
+	// from doing pointless work.
+	if err := r.terminateDupes(pj); err != nil {
+		return nil, fmt.Errorf("terminateDupes failed: %w", err)
+	}
+
+	switch pj.Status.State {
+	case prowv1.PendingState:
+		return nil, r.syncPendingJob(pj)
+	case prowv1.TriggeredState:
+		return r.syncTriggeredJob(pj)
+	case prowv1.AbortedState:
+		return nil, r.syncAbortedJob(pj)
+	}
+
+	return nil, nil
+}
+
+func (r *reconciler) terminateDupes(pj *prowv1.ProwJob) error {
+	pjs := &prowv1.ProwJobList{}
+	if err := r.pjClient.List(r.ctx, pjs, optNotCompletedProwJobs()); err != nil {
+		return fmt.Errorf("failed to list prowjobs: %v", err)
+	}
+
+	return pjutil.TerminateOlderJobs(r.pjClient, r.log, pjs.Items, func(toCancel prowv1.ProwJob) error {
+		client, ok := r.buildClients[pj.ClusterAlias()]
+		if !ok {
+			return fmt.Errorf("no client for cluster %q present", pj.ClusterAlias())
+		}
+		podToDelete := &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: r.config().PodNamespace,
+				Name:      toCancel.Name,
+			},
+		}
+		if err := client.Delete(r.ctx, podToDelete); err != nil {
+			return fmt.Errorf("failed to delete pod %s/%s: %w", podToDelete.Namespace, podToDelete.Name, err)
+		}
+		return nil
+	})
+}
+
+func (r *reconciler) syncPendingJob(pj *prowv1.ProwJob) error {
```

Comment:
Might be good to define what `Pending` (etc) means in the GoDoc for these reconcile methods

Reply:
I can extend it, I actually just copied this code for the most part

Reply:
Added godocs